### PR TITLE
[FLINK-12479][operators] Integrate StreamInputProcessor(s) with mailbox

### PR DESCRIPTION
**Pull request by pnowojski**
This PR makes StreamInputProcessors truly non blocking and integrate availibility/finished states with the mailbox.

This PR depends on https://github.com/apache/flink/pull/8692

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by pnowojski** [comment]
> Suggestion: you can rename old StreamInputProcessor to StreamOneInputProcessor first, in a separate commit. And introduce a StreamInputProcessor interface as follow up commit. This way it should preserve git history of the newly named StreamOneInputProcessor.

I didn't know that it can work like that. Will do :)

> Question: do I get it right that StreamTwoInputProcessor.processInput is still blocking, namely in barrierHandler.isAvailable().get()?

ops. Missed that. I'm not sure if it's worth adding some tests for that, it might be an overkill, especially that this will get a pretty good test coverage once we merge in progress/pending mailbox PRs.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java [hunk 1]
```diff
@@ -195,14 +216,11 @@ public void cleanup() throws Exception {
...
-		throws InterruptedException, ExecutionException, IOException {
-
+	private int selectNextReadingInputIndex() {
+		updateAvailability();
 		int readingInputIndex;
-		while ((readingInputIndex = inputSelection.fairSelectNextIndexOutOf2(availableInputsMask, lastReadInputIndex)) == -1) {
-			if (!waitForAvailableInput(inputSelection)) {
-				return -1;
-			}
+		if ((readingInputIndex = inputSelection.fairSelectNextIndexOutOf2(availableInputsMask, lastReadInputIndex)) == -1) {
```
Comment (1u0) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: Now the `readingInputIndex` doesn't need to be set in a loop. You can change this code to:
```java
int readingInputIndex = inputSelection.fairSelectNextIndexOutOf2(availableInputsMask, lastReadInputIndex);
if (readingInputIndex == -1) { // OR even delegate this check to next `if`
    return -1;
}
...
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java [hunk 2]
```diff
@@ -286,64 +316,33 @@ private void checkAndSetAvailable(int inputIndex) {
 		}
 	}
 
-	/**
-	 * @return false if both of the inputs are finished, true otherwise.
-	 */
-	private boolean waitForAvailableInput(InputSelection inputSelection)
-		throws ExecutionException, InterruptedException, IOException {
-
+	private CompletableFuture<?> waitForAvailableInput(InputSelection inputSelection) {
```
Comment (1u0) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: What about renaming all old `StreamTwoInputSelectable.waitFor...Input(...)` methods?
Now they are not actually waiting, but return "availability" future.
For example, they can be named as `getInputAvailable[Future](...), getOneInputAvailable[Future](...)`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java [hunk 3]
```diff
@@ -286,64 +316,33 @@ private void checkAndSetAvailable(int inputIndex) {
...
-		throws ExecutionException, InterruptedException {
-
-		CompletableFuture<?> future1 = input1.isFinished() ? UNAVAILABLE : input1.isAvailable();
-		CompletableFuture<?> future2 = input2.isFinished() ? UNAVAILABLE : input2.isAvailable();
-
-		if (future1 == UNAVAILABLE && future2 == UNAVAILABLE) {
-			return false;
+	private CompletableFuture<?> waitForAvailableEitherInput() {
+		if (input1.isFinished() && input2.isFinished()) {
+			return AVAILABLE;
```
Comment (1u0) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: I think this should not happen, but if it's just to cover all cases, I'd prefer to return here `UNAVAILABLE`.

Additional note about renaming this method: "either" means **mutually exclusive** (out of two options) in English. I'd name this method something like `getAnyInputAvailableFuture()`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java [hunk 4]
```diff
@@ -286,64 +316,33 @@ private void checkAndSetAvailable(int inputIndex) {
...
+		return CompletableFuture.anyOf(input1.isAvailable(), input2.isAvailable());
 	}
 
-	private void waitForOneInput(StreamTaskInput input)
-		throws IOException, ExecutionException, InterruptedException {
-
+	private CompletableFuture<?> waitForOneInput(StreamTaskInput input) {
 		if (input.isFinished()) {
-			throw new IOException("Could not read the finished input: input" + (input.getInputIndex() + 1) +  ".");
+			throw new IllegalStateException("Could not read the finished input: input" + (input.getInputIndex() + 1) +  ".");
```
Comment (1u0) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: Personally, I'd inline this method.
Otherwise:
 * `Preconditions.checkState`? 😃
 * Same here, suggestion to give it a new name.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 5]
```diff
@@ -246,7 +258,18 @@ protected StreamTask(
...
-	protected abstract void performDefaultAction(ActionContext context) throws Exception;
+	protected void performDefaultAction(ActionContext context) throws Exception {
+		if (!inputProcessor.processInput()) {
+			if (inputProcessor.isFinished()) {
+				context.allActionsCompleted();
+			}
+			else {
+				context.actionsUnavailable();
+				inputProcessor.isAvailable().thenRun(
+					() -> ExceptionUtils.suppressExceptions(context::actionsAvailable));
```
Comment (1u0) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java]: This probably would be addressed by @StefanRRichter's changes. But just in case, what about the following version:
```java
inputProcessor.isAvailable().thenRunAsync(
    () -> ExceptionUtils.suppressExceptions(context::actionsAvailable), workerExecutor);
context.actionsUnavailable();
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java (hunk 4 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: I have also realised, that this shouldn't be a `IllegalStateException`, since this is a user error - I have refactored the code accordingly.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java (hunk 5 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java]: Yes, I've already changed this accordingly to the @StefanRRichter changes.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java [hunk 6]
```diff
@@ -211,6 +211,34 @@ public void setup() throws IOException {
 
 		BufferPool bufferPool = bufferPoolFactory.get();
 		setBufferPool(bufferPool);
+
+		requestPartitions();
+	}
+
+	private void requestPartitions() throws IOException, InterruptedException {
+		synchronized (requestLock) {
+			if (!requestedPartitionsFlag) {
```
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: Have you considered removing the `requestedPartitionsFlag`. It seems like this was just needed for lazy initialization and can potentially go now.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/UnionInputGate.java [hunk 7]
```diff
@@ -141,17 +141,6 @@ public boolean isFinished() {
 		return inputGatesWithRemainingData.isEmpty();
 	}
 
-	@Override
-	public void requestPartitions() throws IOException, InterruptedException {
-		if (!requestedPartitionsFlag) {
```
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/UnionInputGate.java]: Seems like this field can be removed completely from the class now.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTaskNetworkInput.java [hunk 8]
```diff
@@ -148,6 +148,9 @@ public boolean isFinished() {
 
 	@Override
 	public CompletableFuture<?> isAvailable() {
+		if (currentRecordDeserializer != null) {
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTaskNetworkInput.java]: Nit: I would use ternary operator instead of `if` here

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+		initializeNumRecordsIn();
+
+		StreamElement recordOrMark = input.pollNextNullable();
+		if (recordOrMark != null) {
+			int channel = input.getLastChannel();
+			checkState(channel != StreamTaskInput.UNSPECIFIED);
+
+			processElement(recordOrMark, channel);
+		}
+		checkFinished();
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java]: Any reason why we need to check this in both cases and not just `recordOrMark == null`? It uses `synchronized` so saving calls might be good.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java [hunk 10]
```diff
@@ -286,64 +334,17 @@ private void checkAndSetAvailable(int inputIndex) {
...
-	private boolean checkFinished() throws Exception {
-		if (getInput(lastReadInputIndex).isFinished()) {
-			synchronized (lock) {
-				operatorChain.endInput(getInputId(lastReadInputIndex));
-				inputSelection = inputSelector.nextSelection();
-			}
-		}
-
-		return input1.isFinished() && input2.isFinished();
+		return CompletableFuture.anyOf(input1.isAvailable(), input2.isAvailable());
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: You could strongly consider checking if both returned `AVAILABLE` and then also return `AVAILABLE`, and use the `anyOf` only if there is really something to wait for. Some code might only check for the referencial equality and then the method breaks the contract.

Comment (StefanRRichter) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: I would suggest this instead:
<code>
	private CompletableFuture<?> isAnyInputAvailable() {

		if (input1.isFinished()) {
			return input2.isFinished() ? AVAILABLE : input2.isAvailable();
		}

		if (input2.isFinished()) {
			return input1.isAvailable();
		}

		CompletableFuture<?> input1Available = input1.isAvailable();
		CompletableFuture<?> input2Available = input2.isAvailable();

		return (input1Available == AVAILABLE || input2Available == AVAILABLE) ?
			AVAILABLE : CompletableFuture.anyOf(input1Available, input2Available);
	}
<code>

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java [hunk 11]
```diff
@@ -162,31 +176,43 @@ public boolean processInput() throws Exception {
 			if (recordOrMark != null) {
 				processElement1(recordOrMark, input1.getLastChannel());
 			}
+			checkFinished(input1, lastReadInputIndex);
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: Similar question, could this be done only for the `recordOrMark == null` case?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java [hunk 12]
```diff
@@ -162,31 +176,43 @@ public boolean processInput() throws Exception {
...
 		} else {
 			recordOrMark = input2.pollNextNullable();
 			if (recordOrMark != null) {
 				processElement2(recordOrMark, input2.getLastChannel());
 			}
+			checkFinished(input2, lastReadInputIndex);
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: Dito

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java [hunk 13]
```diff
@@ -195,14 +221,12 @@ public void cleanup() throws Exception {
...
-	private int selectNextReadingInputIndex()
-		throws InterruptedException, ExecutionException, IOException {
-
-		int readingInputIndex;
-		while ((readingInputIndex = inputSelection.fairSelectNextIndexOutOf2(availableInputsMask, lastReadInputIndex)) == -1) {
-			if (!waitForAvailableInput(inputSelection)) {
-				return -1;
-			}
+	private int selectNextReadingInputIndex() throws IOException {
+		updateAvailability();
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: Looks weird to me to have this check (and maybe also the one on the next line) in this method. I would call this as part of `processInput` instead, e.g. before this method is called. I know it does the same, but somehow it feels like the check belong into the caller method and not down here.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java [hunk 14]
```diff
@@ -189,51 +189,55 @@ public void testBackwardsEventWithUninitializedChannel() throws Exception {
...
 		// Setup reader with one local and one unknown input channel
 
-		final SingleInputGate inputGate = createInputGate();
-		final BufferPool bufferPool = mock(BufferPool.class);
-		when(bufferPool.getNumberOfRequiredMemorySegments()).thenReturn(2);
+		NettyShuffleEnvironment environment = createNettyShuffleEnvironment();
+		final SingleInputGate inputGate = createInputGate(environment, 2, ResultPartitionType.PIPELINED);
+		try {
+			final BufferPool bufferPool = mock(BufferPool.class);
+			when(bufferPool.getNumberOfRequiredMemorySegments()).thenReturn(2);
```
Comment (1u0) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java]: `bufferPool` is not used here anymore and can be removed.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java [hunk 15]
```diff
@@ -18,213 +18,17 @@
...
- * {@link StreamStatus} events, and forwards them to event subscribers once the
- * {@link StatusWatermarkValve} determines the {@link Watermark} from all inputs has advanced, or
- * that a {@link StreamStatus} needs to be propagated downstream to denote a status change.
- *
- * <p>Forwarding elements, watermarks, or status status elements must be protected by synchronizing
- * on the given lock object. This ensures that we don't call methods on a
- * {@link OneInputStreamOperator} concurrently with the timer callback or other things.
- *
- * @param <IN> The type of the record that can be read with this record reader.
+ * Interface for processing records by StreamTask.
```
Comment (1u0) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java]: Suggestion, use proper class link:
`Interface for processing records by {@link org.apache.flink.streaming.runtime.tasks.StreamTask}.`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java [hunk 16]
```diff
@@ -18,213 +18,17 @@
...
  */
-@Internal
```
Comment (1u0) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java]: Keep the `@Internal` annotation?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java [hunk 17]
```diff
@@ -18,213 +18,17 @@
...
-				}
-			} catch (Exception e) {
-				throw new RuntimeException("Exception occurred while processing valve output stream status: ", e);
-			}
-		}
-	}
+public interface StreamInputProcessor extends AvailabilityListener, Closeable {
+	/**
+	 * @return false if probably there are not more records available and caller should check
+	 * {@link #isAvailable()}. return true if {@link #isAvailable()} check can be safely omitted.
```
Comment (1u0) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java]: I'd rephrase this paragraph. Maybe something like:
```
Return true if more records can be processed.
False, if there are no more records available at the moment and the caller should check {@link #isFinished()} and/or {@link #isAvailable()}.
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 18]
```diff
@@ -243,26 +249,15 @@ protected StreamTask(
...
-					letter.run();
-				}
+	protected void performDefaultAction(DefaultActionContext context) throws Exception {
+		if (!inputProcessor.processInput()) {
+			if (inputProcessor.isFinished()) {
+				context.allActionsCompleted();
+			}
+			else {
+				SuspendedMailboxDefaultAction suspendedDefaultAction = context.suspendDefaultAction();
+				inputProcessor.isAvailable().thenRun(suspendedDefaultAction::resume);
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java]: I would suggest to change this line to `inputProcessor.isAvailable().thenRunAsync(suspendedDefaultAction::resume, getTaskMailboxExecutor());` so that we ensure the resume is executed in the mailbox thread and not by another thread that might get interrupted before it is able to deliver the resume message to the mailbox.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java (hunk 6 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: That might be not that simple, since this flag is also used during `updateInputChannel` from unknown channel. I'm not sure, but update probably can happen concurrently to the setup.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTaskNetworkInput.java (hunk 8 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTaskNetworkInput.java]: Personally I'm not a fan of them, I hope that won't be a review blocker ;)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java (hunk 9 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java]: But it only does it if the input `isFinished()`, so that should be a non issue.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java (hunk 13 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: Hmmm, I would lightly disagree. `updateAvailability()` is (I think) only relevant for `selectNextReadingInputIndex()` so it should be tight to it in case of:
- someones refactor the code and moves `selectNextReadingInputIndex` somewhere else
- someones adds another code path to `selectNextReadingInputIndex`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java (hunk 17 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java]: I've rephrased your rephrased version to:
```
	/**
	 * @return true if {@link StreamInputProcessor} estimates that more records can be processed
	 * immediately. Otherwise false, which means that there are no more records available at the
	 * moment and the caller should check {@link #isFinished()} and/or {@link #isAvailable()}.
	 */
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/execution/MailboxProcessor.java [hunk 19]
```diff
@@ -286,12 +288,20 @@ public void resume() {
 			if (taskMailboxExecutor.isMailboxThread()) {
 				resumeInternal();
 			} else {
-				try {
-					mailbox.putMail(this::resumeInternal);
-				} catch (InterruptedException ie) {
+				boolean delivered = false;
+				boolean interrupted = false;
+				while (!delivered) {
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/execution/MailboxProcessor.java]: With the latest update to my PR, making the mailbox unbounded, this whole fix should no longer be required.

**Comment by pnowojski** [comment]
I have applied the requested changes and rebased the PR on the latest master branch.

**Comment by pnowojski** [comment]
Travis is green, modulo one timeout in end-to-end tests when connecting to apache.org. Merging.
